pkgs/applications/science/misc/root/cling.nix

```diff
@@ -0,0 +1,56 @@
+{ stdenv, fetchurl, python, wget, fetchFromGitHub, libffi, cacert, git, cmake, llvm, ncurses, zlib, fetchgit, glibc, makeWrapper }:
+
+stdenv.mkDerivation rec {
+  name = "cling";
+
+  clingSrc = fetchFromGitHub {
+    owner = "root-project";
+    repo = "cling";
+    rev = "5789f5b2eb14ff8e14e6297fe3a449d0ef794f17";
+    sha256 = "06l0c9p8102prhanhv8xawzq4k8g587d46aykc2y5j7rrzsbs5hs";
+  };
+
+  clangSrc = fetchgit {
+    url = "http://root.cern.ch/git/clang.git";
+    rev = "dd71e0397cfc8667d4c75ef5b4a1b35802608e59"; # This is the head of the cling-patches branch
+    sha256 = "1ir1s4wy9683p9wrvra4xb4xqcd0amqnm4wahx1pf72zk2yxw1wh";
+  };
+
+  llvmSrc = fetchgit {
+    url = "http://root.cern.ch/git/llvm.git";
+    rev = "e0b472e46eb5861570497c2b9efabf96f2d4a485"; # This is the head of the cling-patches branch
+    sha256 = "0yls35vyfcb14wghryss9xsin4cbpgkqckg72czh5jd2w0vjcmbx";
+  };
+
+  srcs = [clingSrc llvmSrc clangSrc];
+
+  unpackCmd = ''
+    mkdir -p ./all
+    if [[ $curSrc == *"clang"* ]]; then
+      cp -r $curSrc ./all/clang
+    elif [[ $curSrc == *"llvm"* ]]; then
+      cp -r $curSrc ./all/llvm
+    else
+      cp -r $curSrc ./all/cling
+    fi
+  '';
```

Comment:
Is this an idiomatic way to have nix put source files in folders?

(Seems like this is an adaption of what was suggested in https://github.com/NixOS/nixpkgs/issues/29661#issuecomment-333632605?)

Reply:
I have no idea, I've never seen another Nix derivation that pulls from multiple repos. I was following the Cling build instructions, which want things laid out a certain way: https://root.cern.ch/cling-build-instructions

Reply:
fcitx does that pkgs/tools/inputmethods/fcitx/unwrapped.nix .
There is also an attribute called `srcs` instead of `src` that you can lookup in pkgs/stdenv/generic/setup.sh . Overriding the unpackCmd is wrong I believe.

Reply:
@teto I already use `srcs` in this derivation. 

It looks like the `fcitx` example you mentioned copies some data (not code) around in the `prePatch` phase. Copying source repos around in the patch phase doesn't seem appropriate though. The manual says "the unpack phase is responsible for unpacking the source code of the package" and that's exactly what I'm doing.

It would be nice to get rid of the custom `unpackPhase`, but going with the default produces the error `unpacker produced multiple directories` (and the directories also don't end up with the desired names).

Reply:
ok sry, I was parsing very fast the comments as having cling in nixpkgs sounded like a great achievement. Good job.

Reply:
Alright, let's leave this like this unless someone more experienced than me objects. (Because I also have no idea what the alternatives would be.)

Reply:
Using this override for `unpackCmd` has one downside that this will not handle tarballs properly. Also it looks only at $srcs, which makes the source overrides slightly less direct:
`cling.overrideAttrs (old: { clingSrc = ./cling; })` will not work, it would require `cling.overrideAttrs (old: rec { clingSrc = ./cling; srcs = [ clingSrc old.llvmSrc old.clangSrc ]; })`
I would just override unpackPhase with something like:
```
unpackPhase = ''
  runHook preUnpack

  cp -r "$llvmSrc" llvm/
  cp -r "$clingSrc" llvm/tools/cling
  cp -r "$clangSrc" llvm/tools/clang

  runHook postUnpack
'';
```
This, however, doesn't solve the tarball issue, which is much harder and not as significant.

Reply:
Hmm, doing that results in a folder like `llvm/cb8mgkfwxrbxjh8srhrqn5rqv5gd0zpf-llvm-e0b472e`. That's why I was doing the bash tests with the asterisks before. @veprbl do you know how to make the source folder have a predictable name?

Reply:
Nvm, silly question. Figured it out, resolving this now

Reply:
Oh hmm @veprbl  I'm not sure it was a good idea to override the entire `unpackPhase`. It results in the source files not being writable, which ends up breaking the build. I think making the source files writable is something the default `unpackPhase` does (controlled by the `dontMakeSourcesWritable` option). 

I had to add a `chmod -R u+w llvm` command to the `unpackPhase` to make it work again which seems a little messy.

Reply:
I don't think you need all sources to be writable. What needs to be writable is just the "build" directory.

Reply:
Though there is nothing wrong with doing "chmod -R u+w llvm". We prefer to have a writable copy of sources to be able to add custom patches.

Reply:
It looked like the build process was copying some random file into the build directory and then trying to modify it. I'll keep the full `chmod` for now. Resolving this...